pkg/ruler/ruler.go

```diff
@@ -507,8 +530,23 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRu
 			}
 		}
 	} else {
-		// Even if the replication factor is set to a number bigger than 1, only the first ruler evaluates the rule group
-		ownsRuleGroup = rlrs.Instances[0].Addr == instanceAddr
+		if r.Config().EnableHAEvaluation {
+			for i, ruler := range rlrs.Instances {
+				if ruler.Addr == instanceAddr && i == 0 {
+					level.Debug(r.Logger()).Log("msg", "primary taking ownership", "user", g.User, "group", g.Name, "namespace", g.Namespace, "ruler", instanceAddr)
+					ownsRuleGroup = true
+					break
+				}
+				if ruler.Addr == instanceAddr && r.nonPrimaryInstanceOwnsRuleGroup(g, rlrs.GetAddresses()[:i], instanceAddr) {
+					level.Info(r.Logger()).Log("msg", "non-primary ruler taking ownership", "user", g.User, "group", g.Name, "namespace", g.Namespace, "ruler", instanceAddr)
+					ownsRuleGroup = true
+					break
+				}
+			}
+		} else {
+			// Even if the replication factor is set to a number bigger than 1, only the first ruler evaluates the rule group
+			ownsRuleGroup = rlrs.Instances[0].Addr == instanceAddr
+		}
 	}
 
 	if ownsRuleGroup && ruleGroupDisabled(g, disabledRuleGroups) {
```

Comment:
nit: we could create a function to generate the return in case of ownsRuleGroup = true.

this way we can retur directly from the code with early returns instead of using this nesting which is confusing and hard to reason about.

```golang
func (r *Ruler) instanceOwnsRuleGroup(rr ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRuleGroups validation.DisabledRuleGroups, instanceAddr string, forBackup bool) (bool, error) {

	hash := tokenForGroup(g)

	rlrs, err := rr.Get(hash, RingOp, nil, nil, nil)
	if err != nil {
		return false, errors.Wrap(err, "error reading ring to verify rule group ownership")
	}

	if forBackup {
		// Only the second up to the last replica are used as backup
		for i := 1; i < len(rlrs.Instances); i++ {
			if rlrs.Instances[i].Addr == instanceAddr {
				return ownsRuleGroupOrDisable(g, disabledRuleGroups)
			}
		}
	} 
        if rlrs.Instances[0].Addr == instanceAddr {
              // regardless of ruler HA is enabled or not, in this case this ruler is the primary
               return ownsRuleGroupOrDisable(g, disabledRuleGroups)
        } 
	if r.Config().EnableHAEvaluation {
		for i, ruler := range rlrs.Instances {
			if  i == 0 {
                                 // we already checked for primary
                                  continue
			}
			if ruler.Addr == instanceAddr && r.nonPrimaryInstanceOwnsRuleGroup(g, rlrs.GetAddresses()[:i], instanceAddr) {
				level.Info(r.Logger()).Log("msg", "non-primary ruler taking ownership", "user", g.User, "group", g.Name, "namespace", g.Namespace, "ruler", instanceAddr)
				return ownsRuleGroupOrDisable(g, disabledRuleGroups)
			}
		}		
	} 
        // means for sure that the instance is not the owner
	return false, nil
}

func ownsRuleGroupOrDisable(g *rulespb.RuleGroupDesc, disabledRuleGroups validation.DisabledRuleGroups)
	if ruleGroupDisabled(g, disabledRuleGroups) {
		return false, &DisabledRuleGroupErr{Message: fmt.Sprintf("rule group %s, namespace %s, user %s is disabled", g.Name, g.Namespace, g.User)}
	}
        return true, nil
}
```

Reply:
done